protocols/autonatv2/src/client/behaviour.rs

```diff
@@ -0,0 +1,228 @@
+use std::{
+    collections::{HashSet, VecDeque},
+    sync::Arc,
+    task::{Context, Poll},
+};
+
+use ip_global::IpExt;
+use libp2p_core::{multiaddr::Protocol, transport::PortUse, Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    dial_opts::DialOpts, ConnectionDenied, ConnectionHandler, ConnectionId, FromSwarm,
+    NetworkBehaviour, NotifyHandler, ToSwarm,
+};
+use rand_core::RngCore;
+
+use crate::{
+    client::{ResponseInfo, ToBehaviour},
+    generated::structs::{mod_DialResponse::ResponseStatus, DialStatus},
+    request_response::{DialRequest, DialResponse},
+};
+
+use super::{handler::Handler, FromBehaviour};
+
+enum Command {
+    TestListenerReachability {
+        server_peer: PeerId,
+        local_addrs: Vec<Multiaddr>,
+    },
+}
+
+pub struct ReachabilityTestSucc {
+    pub server_peer: PeerId,
+    pub visible_addr: Option<Multiaddr>,
+    pub suspicious_addrs: Vec<Multiaddr>,
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ReachabilityTestErr {
+    #[error("Server chose not to dial any provided address. Server peer id: {peer_id}")]
+    ServerChoseNotToDialAnyAddress { peer_id: PeerId },
+    #[error("Server rejected dial request. Server peer id: {peer_id}")]
+    ServerRejectedDialRequest { peer_id: PeerId },
+    #[error("Server ran into an internal error. Server peer id: {peer_id}")]
+    InternalServerError { peer_id: PeerId },
+    #[error("Server did not respond correctly to dial request. Server peer id: {peer_id}")]
+    InvalidResponse { peer_id: PeerId },
+    #[error("Server was unable to connect to address: {addr:?}. Server peer id: {peer_id}")]
+    UnableToConnectOnSelectedAddress {
+        peer_id: PeerId,
+        addr: Option<Multiaddr>,
+    },
+    #[error("Server experienced failure during dial back on address: {addr:?} Server peer id: {peer_id}")]
+    FailureDuringDialBack {
+        peer_id: PeerId,
+        addr: Option<Multiaddr>,
+    },
+}
+
+pub enum Event {
+    CompletedReachabilityTest(Result<ReachabilityTestSucc, ReachabilityTestErr>),
+}
```

Comment:
In general, I'd like to be conservative with the amount of information that we expose in our events. Anything that is public API will be a breaking change to change later! :)

For example, in `libp2p-dcutr`, I did a refactoring recently to change the API to essentially just return a `Result`: https://github.com/libp2p/rust-libp2p/pull/4749/files#diff-e673e0cdaa390b5f700c2fcf9a99b3c953d37b990c6c36dfaa078178def9fa59

For the client, this could be:

```rust
struct Event {
	/// The ID of the Autonat Server.
	server: PeerId,
	/// The addresses that was tested for reachability.
	candidate: Multiaddr,
	/// The outcome of the test.
	result: Result<(), Error> // `Error` should probably be opaque.
}
```

For the server, I think something similar makes sense where we report a single event per probe with the ID of the client, the address to be tested and the outcome.

Reply:
I think we addressed that. Right?